File: cpp-linux/annotations/highlight_annotations/highlight_annotations.cpp
// highlight_annotations -- C++ port of examples\Vb6\annotations\highlight_annotations\highlight_annotations.bas
// Highlight / squiggly / strikeout / underline annotations over text.
#include <lumaspdf.h>
#include <cstdio>
#include <string>

static const UI32 clYellow = 65535;
static const UI32 clRed = 255;

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    return 0;
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, nullptr, PDFError);
    pdfCreateNewPDFA(pdf, "");

    pdfSetPageCoords(pdf, pcTopDown);

    pdfAppend(pdf);
    const char* text = "Some text on a page...";
    pdfSetFontA(pdf, "Helvetica", fsRegular, 20.0, 0, cp1252);

    double d = pdfGetDescent(pdf);
    double w = pdfGetTextWidthA(pdf, text);

    pdfWriteTextA(pdf, 50.0, 50.0, text);
    pdfHighlightAnnotA(pdf, atHighlight, 50.0, 50.0 + d, w, 20.0, clYellow, "Test app", "Highligh Annotations", "This is a highlight annotation");

    pdfWriteTextA(pdf, 50.0, 80.0, text);
    pdfHighlightAnnotA(pdf, atSquiggly, 50.0, 80.0 + d, w, 20.0, clRed, "Test app", "Squiggly Annotations", "This is a squiggly annotation");

    pdfWriteTextA(pdf, 50.0, 110.0, text);
    pdfHighlightAnnotA(pdf, atStrikeOut, 50.0, 110.0 + d, w, 20.0, clRed, "Test app", "Strikeout Annotations", "This is a strikeout annotation");

    pdfWriteTextA(pdf, 50.0, 140.0, text);
    pdfHighlightAnnotA(pdf, atUnderline, 50.0, 140.0 + d, w, 20.0, clRed, "Test app", "Underline Annotations", "This is a underline annotation");
    pdfEndPage(pdf);

    if(pdfHaveOpenDoc(pdf) != 0){
        std::string outFile = exeDir(argv[0]) + "/out.pdf";
        if(pdfOpenOutputFileA(pdf, outFile.c_str()) == 0){
            pdfDeletePDF(pdf);
            return 0;
        }
        if(pdfCloseFile(pdf) != 0){
            printf("PDF file \"%s\" successfully created!\n", outFile.c_str());
        }
    }

    pdfDeletePDF(pdf);
    return 0;
}

File: python/annotations/highlight_annotations/highlight_annotations.py
# ============================================================================
#  highlight_annotations -- Python (ctypes) port of
#  examples\Vb6\annotations\highlight_annotations\highlight_annotations.bas
#  Highlight / squiggly / strikeout / underline annotations over text.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

clYellow = 65535
clRed = 255


@L.TErrorProc
def err_proc(data, errcode, errmsg, errtype):
    return 0


def main():
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, err_proc)
    L.pdfCreateNewPDFA(pdf, b"")

    L.pdfSetPageCoords(pdf, L.pcTopDown)

    L.pdfAppend(pdf)
    text = b"Some text on a page..."
    L.pdfSetFontA(pdf, b"Helvetica", L.fsRegular, 20.0, 0, L.cp1252)

    d = L.pdfGetDescent(pdf)
    w = L.pdfGetTextWidthA(pdf, text)

    L.pdfWriteTextA(pdf, 50.0, 50.0, text)
    L.pdfHighlightAnnotA(pdf, L.atHighlight, 50.0, 50.0 + d, w, 20.0, clYellow, b"Test app", b"Highligh Annotations", b"This is a highlight annotation")

    L.pdfWriteTextA(pdf, 50.0, 80.0, text)
    L.pdfHighlightAnnotA(pdf, L.atSquiggly, 50.0, 80.0 + d, w, 20.0, clRed, b"Test app", b"Squiggly Annotations", b"This is a squiggly annotation")

    L.pdfWriteTextA(pdf, 50.0, 110.0, text)
    L.pdfHighlightAnnotA(pdf, L.atStrikeOut, 50.0, 110.0 + d, w, 20.0, clRed, b"Test app", b"Strikeout Annotations", b"This is a strikeout annotation")

    L.pdfWriteTextA(pdf, 50.0, 140.0, text)
    L.pdfHighlightAnnotA(pdf, L.atUnderline, 50.0, 140.0 + d, w, 20.0, clRed, b"Test app", b"Underline Annotations", b"This is a underline annotation")
    L.pdfEndPage(pdf)

    if L.pdfHaveOpenDoc(pdf) != 0:
        out_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "out.pdf")
        if L.pdfOpenOutputFileA(pdf, out_file.encode("latin-1")) == 0:
            L.pdfDeletePDF(pdf)
            return
        if L.pdfCloseFile(pdf) != 0:
            print('PDF file "' + out_file + '" successfully created!')

    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()
